shells/android/javatests/arcs/storage/ServiceStoreTest.kt

```diff
@@ -0,0 +1,153 @@
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.storage
+
+import android.content.ComponentName
+import android.content.ServiceConnection
+import androidx.lifecycle.Lifecycle
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import arcs.crdt.CrdtCount
+import arcs.crdt.parcelables.ParcelableCrdtType
+import arcs.data.CountType
+import arcs.storage.driver.RamDisk
+import arcs.storage.driver.RamDiskDriverProvider
+import arcs.storage.driver.RamDiskStorageKey
+import arcs.storage.parcelables.ParcelableStoreOptions
+import arcs.storage.parcelables.toParcelable
+import arcs.storage.service.BindingContext
+import arcs.storage.service.ConnectionFactory
+import arcs.storage.service.DeferredResult
+import arcs.storage.service.StorageServiceBindingDelegate
+import arcs.storage.service.StorageServiceConnection
+import com.google.common.truth.Truth.assertThat
+import com.nhaarman.mockitokotlin2.mock
+import kotlin.coroutines.coroutineContext
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.yield
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/** Tests for the [ServiceStore]. */
+@RunWith(AndroidJUnit4::class)
+@UseExperimental(ExperimentalCoroutinesApi::class)
+class ServiceStoreTest {
+    private lateinit var lifecycle: Lifecycle
+    private val storeOptions = StoreOptions<CrdtCount.Data, CrdtCount.Operation, Int>(
+        RamDiskStorageKey("myData"),
+        ExistenceCriteria.ShouldCreate,
+        CountType()
+    )
+
+    @Before
+    fun setup() {
+        RamDisk.clear()
+        RamDiskDriverProvider()
+        lifecycle = mock()
+    }
+
+    @Test
+    fun getLocalData_getsLocalDataFromService() = runBlocking {
+        val service = buildService(storeOptions.toParcelable(ParcelableCrdtType.Count))
+        val bindingDelegate = buildBindingDelegate(service)
+        val connectionFactory = buildConnectionFactory(bindingDelegate)
+
+        val store = ServiceStore(
+            storeOptions,
+            ParcelableCrdtType.Count,
+            lifecycle,
+            connectionFactory,
+            coroutineContext
+        ).initialize()
+
+        val data = store.getLocalData()
+
+        assertThat(data.values).isEmpty()
+
+        // Increment at the binding context directly.
+        val deferredResult = DeferredResult(coroutineContext)
+        service.sendProxyMessage(
+            ProxyMessage.Operations<CrdtCount.Data, CrdtCount.Operation, Int>(
+                listOf(CrdtCount.Operation.Increment("alice", 0 to 1)),
+                null
+            ).toParcelable(ParcelableCrdtType.Count),
+            deferredResult
+        )
+
+        deferredResult.await()
+
+        val dataAfter = store.getLocalData()
+
+        assertThat(dataAfter.values).containsExactly("alice", 1).inOrder()
+
+        store.onLifecycleDestroyed()
```

Comment:
Could put this in a `tearDown` method, maybe?

Reply:
in order for runBlocking to work correctly, it's important that all of the coroutines launched from that context are finished before the runblocking can finish, which makes it infeasible to put `onLifecycleDestroyed` in a tearDown method (since that would be called from outside the `runBlocking`, unfortunately.